scenario/smartsearch.go

```diff
@@ -0,0 +1,200 @@
+	var warnings []string
+	if len(settings.searchTerms) == 0 {
+		return warnings, errors.New("no search terms")
+	}
+	return warnings, nil
+}
+
+func intPtr(i int) *int {
+	return &i
+}
+
+func quote(s string) string {
+	return fmt.Sprintf(`"%s"`, s)
+}
+
+func ternary(cond bool, ifCondTrue string, ifCondFalse string) string {
+	if cond {
+		return ifCondTrue
+	}
+	return ifCondFalse
+
+}
+
+func logSmartSearchResults(logEntry *logger.LogEntry, searchResult *enigma.SearchResult, searchSuggestionResult *enigma.SearchSuggestionResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	suggestions := []string{}
+	for _, suggestion := range searchSuggestionResult.Suggestions {
+		suggestions = append(suggestions, quote(suggestion.Value))
+	}
+	if len(suggestions) > 0 {
+		logEntry.LogDebugf("search suggestions: %s", strings.Join(suggestions, ","))
+	}
+	for _, sga := range searchResult.SearchGroupArray {
+		searchTermsMatched := []string{}
+		for _, stIndex := range sga.SearchTermsMatched {
+			searchTermsMatched = append(searchTermsMatched, quote(searchResult.SearchTerms[stIndex]))
+		}
+		for _, item := range sga.Items {
+			matches := []string{}
+			for _, match := range item.ItemMatches {
+				matches = append(matches, quote(match.Text))
+			}
+			logEntry.LogDebugf(`search result for term%s %s in %s %s: %s`,
+				ternary(len(searchTermsMatched) != 1, "s", ""), strings.Join(searchTermsMatched, ","),
+				item.ItemType, quote(item.Identifier), strings.Join(matches, ","))
+		}
+	}
+}
+
+var spaceRegex = regexp.MustCompile(`\s+`)
+
+func parseSearchTerms(s string) []string {
+	searchTerms := []string{}
+	sb := &strings.Builder{}
+	quoted := false
+	escaped := false
+	addSearchTerm := func() {
+		defer sb.Reset()
+		searchTerm := sb.String()
+		searchTerm = strings.TrimSpace(searchTerm)
+		searchTerm = spaceRegex.ReplaceAllString(searchTerm, " ")
+		if searchTerm != "" {
+			searchTerms = append(searchTerms, searchTerm)
+		}
+	}
+	for _, r := range s {
+		if escaped {
+			sb.WriteRune(r)
+		} else if r == '\\' {
+			escaped = true
+			continue
+		} else if r == '"' {
+			quoted = !quoted
+		} else if !quoted && unicode.IsSpace(r) {
+			addSearchTerm()
+		} else {
+			sb.WriteRune(r)
+		}
+		escaped = false
+	}
+	addSearchTerm()
+	return searchTerms
+}
+
+// Execute implements ActionSettings interface
+func (settings SmartSearchSettings) Execute(sessionState *session.State, actionState *action.State, connection *connection.ConnectionSettings, label string, reset func()) {
+	uplink := sessionState.Connection.Sense()
+	if uplink.CurrentApp == nil {
+		actionState.AddErrors(errors.New("not connected to app"))
+		return
+	}
+	doc := uplink.CurrentApp.Doc
+	searchTerms := settings.searchTerms
+	sessionState.LogEntry.LogDebugf(`search terms: "%s"`, strings.Join(searchTerms, `","`))
+
+	var searchResult *enigma.SearchResult
+	var searchSuggestionResult *enigma.SearchSuggestionResult
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
```

Comment:
if there's a panic entire gopherciser (or other service using this as lib) would be taken down due to no panic recover. Try to use the async functions functions with callbacks instead followed by a sessionState.Wait (appliest to all the usuage of `go` directly in the file)

Reply:
Fixed